### PR TITLE
Metadata file import controller

### LINKED ISSUE #368: Project Sample Metadata - File Import Controller
## Summary

Develop a `FileImportController` which for now has the `create` action. This new controller will use the Samples::Metadata::FileImportService from https://github.com/phac-nml/irida-next/issues/356

## Acceptance Criteria

- Create a `FileImportController` which will be nested under `projects -> samples -> metadata`
- The strong params for the controller should include `file`, `sample_id_column`, and `ignore_empty_values`
- The controller will utilize the `Samples::Metadata::FileImport` service to validate, parse, and call the metadata update service
- Controller should return a turbo stream response
- Ensure the routes are setup correctly

### PR DESCRIPTION
**Pull request by ksierks**
## What does this PR do and why?
The new controller that calls the Samples::Metadata::FileImportService.
Fixes #368.

## Screenshots or screen recordings
N/A

## How to set up and validate locally
Verify the tests within `test/controllers/projects/samples/metadata/file_import_controller_test.rb` pass.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [ ] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: test/controllers/projects/samples/metadata/file_import_controller_test.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    module Metadata
+      class FileImportControllerTest < ActionDispatch::IntegrationTest
+        setup do
+          sign_in users(:john_doe)
+          @namespace = groups(:group_one)
+          @project = projects(:project1)
+          @csv = fixture_file_upload('test/fixtures/files/metadata/valid.csv')
+        end
+
+        # bin/rails test test/controllers/projects/samples/metadata/file_import_controller_test.rb
```
[the remark this answers is not in the document]
Comment (ksierks) [reply test/controllers/projects/samples/metadata/file_import_controller_test.rb]: f0aec8a5bdba53876052ea00512ca835b9590911

File: config/routes/project.rb [hunk 2]
```diff
@@ -23,6 +23,12 @@
         scope module: :samples, as: :samples do
           collection do
             resource :transfer, only: %i[create new]
+
+            resource :file_import, module: :metadata, only: %i[create]
+
+            # scope module: :metadata, as: :metadata do
```
[the remark this answers is not in the document]
Comment (ksierks) [reply config/routes/project.rb]: f0aec8a5bdba53876052ea00512ca835b9590911

File: config/locales/en.yml [hunk 3]
```diff
@@ -873,6 +873,11 @@ en:
         create:
           success: Samples were successfully transferred.
           error: "The following list of samples failed to transfer:"
+      metadata:
+        file_imports:
+          create:
+            success: File was successfully imported.
+            error: "The following list of samples failed:"
```
[the remark this answers is not in the document]
Comment (ksierks) [reply config/locales/en.yml]: f0aec8a5bdba53876052ea00512ca835b9590911

File: app/controllers/projects/samples/metadata/file_imports_controller.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+          authorize! @project, to: :update_sample?
+          @imported_metadata = ::Samples::Metadata::FileImportService.new(@project, current_user,
+                                                                          file_import_params).execute
+          if @project.errors.empty?
+            render status: :ok, locals: { type: :success, message: t('.success') }
+          elsif @project.errors.include?(:sample)
+            errors = @project.errors.messages_for(:sample)
+            render status: :partial_content, locals: { type: :alert, message: t('.error'), errors: }
+          else
+            error = @project.errors.full_messages_for(:base).first
```
Comment (deepsidhu85) [inline app/controllers/projects/samples/metadata/file_imports_controller.rb:20]: Since we can have multiple errors attached we don't want to just get the first one. I had recently just updated the `transfers_controller` in a similar way. Take a look at the `create` method in `app/controllers/projects/samples/transfers_controller.rb`

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This looks good to me. Just one last comment

File: app/controllers/projects/samples/metadata/file_imports_controller.rb (hunk 4 above)
Comment (ksierks) [reply app/controllers/projects/samples/metadata/file_imports_controller.rb:20]: I can do that but, I don't think you will have more that one error on base.
There are 2 kinds of errors, file validation errors and those thrown from actually updating the metadata. I believe only the first file validation failure throws a SampleMetadataFileImportError, attaches to base, and returns. Therefore, I don't think it matters if I grab all or the first error for base. 
Only the errors attached to sample could have more that one. If a sample fails updating the metadata, it attaches the error and continues to the next sample.
Actually, I was planning to show base and sample errors differently in my UI wire-frames. 
Can you think of a scenario I might have multiple base errors? I can add a new test for that.

Comment (deepsidhu85) [reply app/controllers/projects/samples/metadata/file_imports_controller.rb:20]: I completely misread that, but yeah looking over the service it will only have one error message attached here. Disregard

File: test/controllers/projects/samples/metadata/file_import_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+               params: {
+                 file_import: {
+                   file: csv,
+                   sample_id_column: 'sample_name'
+                 }
+               }
+
+          assert_response :partial_content
+        end
+      end
```
Comment (deepsidhu85) [inline test/controllers/projects/samples/metadata/file_import_controller_test.rb:67]: Can you add in tests for `duplicate headers`, `missing sample id column`, `missing metadata column`, `missing metadata row`, empty file (no file passed). I see you have all the files in the fixtures already

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Everything looks good just a comment about adding some tests

File: test/controllers/projects/samples/metadata/file_import_controller_test.rb (hunk 5 above)
Comment (ksierks) [reply test/controllers/projects/samples/metadata/file_import_controller_test.rb:111]: f2537ec18afa5061edec239fc04fa50ee20d0d01

**Review by deepsidhu85 — APPROVED** [review]

**Review by ericenns — APPROVED** [review]
This looks good to me! Thanks
